cli/src/semdep/find_lockfiles.py

```diff
@@ -110,3 +114,30 @@ def make_dependency_trie(
                 deps = list(parse_lockfile_str(lockfile.read_text(), lockfile))
                 dep_trie.insert(lockfile, deps, namespace)
     return dep_trie
+
+
+@lru_cache(maxsize=None)
+def search_path(p: Path, lockfile_pattern: str) -> Optional[Path]:
+    lockfiles = list(p.glob(lockfile_pattern))
+    if len(lockfiles) > 1:
+        raise SemgrepError(f"target with multiple lockfiles?")
+    elif len(lockfiles) == 1:
+        return lockfiles[0]
+    else:
+        return None
+
+
+def find_single_lockfile(
+    p: Path, ecosystem: PackageManagers
+) -> Optional[Tuple[Path, List[LockfileDependency]]]:
+    parents = list(p.parents)
+    for path in parents:
+        for lockfile_pattern in NAMESPACE_TO_LOCKFILES[ecosystem]:
+            lockfile = search_path(path, lockfile_pattern)
+            if lockfile:
+                return lockfile, list(
+                    parse_lockfile_str(lockfile.read_text(encoding="utf8"), lockfile)
```

Comment:
is utf-8 a fair assumption for lockfiles?

Reply:
Not actually sure

Reply:
gonna go with yeah until it doesn't work lol